pkg/server/backend_manager.go

```diff
@@ -196,176 +192,105 @@ func NewBackend(conn agent.AgentService_ConnectServer) (*Backend, error) {
-	dbm.addBackend(agentID, header.UID, backend)
-}
-
-func (dbm *DefaultBackendManager) RemoveBackend(backend *Backend) {
-	agentID := backend.GetAgentID()
-	klog.V(5).InfoS("Remove the agent from the DefaultBackendManager", "agentID", agentID)
-	dbm.removeBackend(agentID, header.UID, backend)
-}
-
-// DefaultBackendStorage is the default backend storage.
-type DefaultBackendStorage struct {
-	mu sync.RWMutex //protects the following
-	// A map between agentID and its grpc connections.
-	// For a given agent, ProxyServer prefers backends[agentID][0] to send
-	// traffic, because backends[agentID][1:] are more likely to be closed
-	// by the agent to deduplicate connections to the same server.
-	//
-	// TODO: fix documentation. This is not always agentID, e.g. in
-	// the case of DestHostBackendManager.
-	backends map[string][]*Backend
-	// agentID is tracked in this slice to enable randomly picking an
-	// agentID in the Backend() method. There is no reliable way to
-	// randomly pick a key from a map (in this case, the backends) in
-	// Golang.
-	agentIDs []string
-	random   *rand.Rand
-	// idTypes contains the valid identifier types for this
-	// DefaultBackendStorage. The DefaultBackendStorage may only tolerate certain
-	// types of identifiers when associating to a specific BackendManager,
-	// e.g., when associating to the DestHostBackendManager, it can only use the
-	// identifiers of types, IPv4, IPv6 and Host.
-	idTypes []header.IdentifierType
+	proxyStrategies []ProxyStrategy
+
+	// All backends by agentID.
+	all BackendStorage
+	// All backends by host identifier(s). Only used with ProxyStrategyDestHost.
+	byHost BackendStorage
+	// All default-route backends, by agentID. Only used with ProxyStrategyDefaultRoute.
+	byDefaultRoute BackendStorage
 }
 
 // NewDefaultBackendManager returns a DefaultBackendManager.
-func NewDefaultBackendManager() *DefaultBackendManager {
+func NewDefaultBackendManager(proxyStrategies []ProxyStrategy) *DefaultBackendManager {
+	metrics.Metrics.SetBackendCount(0)
 	return &DefaultBackendManager{
-		DefaultBackendStorage: NewDefaultBackendStorage(
-			[]header.IdentifierType{header.UID})}
+		proxyStrategies: proxyStrategies,
+		all:             NewDefaultBackendStorage(),
+		byHost:          NewDefaultBackendStorage(),
+		byDefaultRoute:  NewDefaultBackendStorage(),
+	}
 }
 
-// NewDefaultBackendStorage returns a DefaultBackendStorage
-func NewDefaultBackendStorage(idTypes []header.IdentifierType) *DefaultBackendStorage {
-	// Set an explicit value, so that the metric is emitted even when
-	// no agent ever successfully connects.
-	metrics.Metrics.SetBackendCount(0)
-	return &DefaultBackendStorage{
-		backends: make(map[string][]*Backend),
-		random:   rand.New(rand.NewSource(time.Now().UnixNano())),
-		idTypes:  idTypes,
-	} /* #nosec G404 */
+func (s *DefaultBackendManager) Backend(addr string) (*Backend, error) {
+	for _, strategy := range s.proxyStrategies {
```

Comment:
Why do we want to consolidate all possible proxy strategies into a single Manager?

Reply:
> Why do we want to consolidate all possible proxy strategies into a single Manager?

At HEAD we have something of a mess, with 1:1 ProxyStrategy to BackendManager / BackendStorage, but logic spread between pkg/server/.*backend_manager.go and pkg/server/server.go. This change improves separation of concerns. For example, the main loop (pkg/server/server.go) does not need to be aware of proxy strategies.

After this consolidation, I can make the (single) BackendManager aware of agent drained state more safely.